### DISCUSSION TITLE
Runtime types for method/function overloading

### DISCUSSION THREAD

**Discussion by marcus-sa** [marcj/deepkit#523] [Ideas]
```ts
function method(arg1: number): void;
function method(arg1: string): void;
function method(arg1: any): void {}
```
The example above doesn't cause errors, but the following code is generated
```ts
method.__type = ['arg1', 'method', 'P\'2!$/"'];
method.__type = ['arg1', 'method', 'P&2!$/"'];
function method(arg1) { }
method.__type = ['arg1', 'method', 'P"2!$/"'];
```

We need to come up with a way to have runtime types for method/function overloading.

**Comment by marcus-sa** [comment]
Perhaps we could generate an array of packs for the overloads:
```ts
function method(arg1) { }
method.__type = [['arg1', 'method', 'P\'2!$/"'], ['arg1', 'method', 'P&2!$/"']];
```
But how we would represent this to the user is another thing...

**Reply by marcj** [reply]
when we think about a solution to this we might be able to find a way to make it also work with interface merging

**Reply by marcus-sa** [reply]
@marcj I just found another case where I need runtime types for method/function overloading
https://github.com/marcus-sa/deepkit-restate/blob/b25abbc90674aec2857856fbcbd77c14dcf2269e/src/saga/step-builder.ts#L176-L195
